src/ussd_optout_rapidpro.js

```diff
@@ -1,29 +1,333 @@
 
     var GoNDOH = App.extend(function(self) {
         App.call(self, "state_start");
         var $ = self.$;
 
         self.init = function() {
+            self.rapidpro = new go.RapidPro(
+                new JsonApi(self.im, {headers: {'User-Agent': ["Jsbox/NDoH-Optout"]}}),
+                self.im.config.services.rapidpro.base_url,
+                self.im.config.services.rapidpro.token
+            );
         };
 
-        self.states.add("state_start", function(name) {
-            return new ChoiceState(name, {
-                question: $("Please let us know why you do not want MomConnect messages"),
+        self.contact_in_group = function(contact, groups){
+            var contact_groupids = _.map(_.get(contact, "groups", []), "uuid");
+            return _.intersection(contact_groupids, groups).length > 0;
+        };
+
+        self.add = function(name, creator) {
+            self.states.add(name, function(name, opts) {
+                if (self.im.msg.session_event !== 'new')
+                    return creator(name, opts);
+
+                var timeout_opts = opts || {};
+                timeout_opts.name = name;
+                return self.states.create('state_timed_out', timeout_opts);
+            });
+        };
+
+        self.states.add('state_timed_out', function(name, creator_opts) {
+            return new MenuState(name, {
+                question: $('Welcome back. Please select an option:'),
+                choices: [
+                    new Choice(creator_opts.name, $('Continue signing up for messages')),
+                    new Choice('state_start', $('Main menu'))
+                ]
+            });
+        });
+
+        self.states.add("state_start", function(name, opts) {
+            // Reset user answers when restarting the app
+            self.im.user.answers = {};
+            var msisdn = utils.normalize_msisdn(self.im.user.addr, "ZA");
+
+            return self.rapidpro.get_contact({urn: "tel:" + msisdn})
+                .then(function(contact) {
+                    self.im.user.set_answer("contact", contact);
+                    // Set the language if we have it
+                    if(_.isString(_.get(contact, "language"))) {
+                        return self.im.user.set_lang(contact.language);
+                    }
+                }).then(function() {
+                    // Delegate to the correct state depending on group membership
+                    var contact = self.im.user.get_answer("contact");
+                    if(self.contact_in_group(contact, self.im.config.public_group_ids)) {
+                        return self.states.create("state_opt_out");
+                    } else if(self.contact_in_group(contact, self.im.config.clinic_group_ids)){
+                        return self.states.create("state_opt_out");
+                    } else {
+                        return self.states.create("state_check_previous_optout");
+                    }
```

Comment:
I'm wondering if it might be better to have all of the logic in one place, so you can easily see where the user will end up, without having to check multiple states. So adding something like:
```
if(self.contact_in_group(contact, _.concat(self.im.config.public_group_ids, self.im.config.clinic_group_ids))){
    return self.states.create("state_opt_out");
} else if(self.contact_in_group(contact, self.im.config.optout_group_ids && _.get(contact, "fields.optout_reason") === "")) {
    return self.states.create("state_get_optout_reason");
} else {
    return self.states.create("state_no_previous_optout");
}
```
Where `state_get_optout_reason` is just `state_check_previous_optout` without the if/else logic.

Reply:
will use this suggestion, thank you